src/steps/json_step.go

```diff
@@ -0,0 +1,119 @@
+package steps
+
+import (
+	"encoding/json"
+	"log"
+	"reflect"
+)
+
+// JSONStep is used to store a step in JSON
+type JSONStep struct {
+	Step Step
+}
+
+// MarshalJSON marshals the step to JSON
+func (j *JSONStep) MarshalJSON() (b []byte, e error) {
```

Comment:
Sorry for the second round of comments, I had to sleep over this in order to think this through. 
Doing this in an OO way feels strange to me. What about just using functions here?

```go
package json

func MarshalStep(s Step) (b []byte, e error) {...}
func UnmarshalStep(b []byte) (s Step, err error) {...}
func MarshalStepList(sl StepList) (b []byte, e error) {...}
func UnmarshalStepList(b []byte) (sl StepList, err error) {...}
```

Then you can call this quite nicely:

```go
json, err := json.MarshalStep(currentStep)
step, err := json.UnmarshalStep(fileContent)
json, err := json.MarshalStepList(stepList)
stepList, err := json.UnmarshalStepList(fileContent)
```

From how I understand Go, this seems a bit more idiomatic.
It also nicely separates the different concerns. Now `Step` and `StepList` are just simple POGO (plain old Go objects) which implement only business logic. The logic to serialize them into a config file format is cleanly extracted into a separate layer. I could also see this being a part of the `config` package if you don't want to create a separate `json` package here.

Reply:
I don't think that would compose well. In the end, all I need is the ability to marshal a run state. The run state contains step lists, which in turn contain steps.

If `MarshalStepList` were to make use of `MarshalStep` then I would need to be manually joining the JSON result of `MarshalStep`. 

The current approach makes use of how go marshals / unmarshals any objects. Previously we converted the runstate to a whole new struct and then used go's built in way to marshal / unmarshal. This is minimizing the amount of new structs we need to define

Articles used for reference (although did not follow either fully)
https://blog.gopheracademy.com/advent-2016/advanced-encoding-decoding/
http://gregtrowbridge.com/golang-json-serialization-with-interfaces/

Reply:
Ah, okay, thanks for the clarification! You are right, if this automatically composes, that's even better.